packages/timeline-state-resolver/src/service/stateHandler.ts

```diff
@@ -0,0 +1,220 @@
+	private currentState: ExecutedStateChange<DeviceState, Command> | undefined = undefined
+	private _executingStateChange = false
+
+	private clock: NodeJS.Timeout
+
+	private convertTimelineStateToDeviceState: BaseDeviceAPI<DeviceState, Command>['convertTimelineStateToDeviceState']
+	private diffDeviceStates: BaseDeviceAPI<DeviceState, Command>['diffStates']
+	private executeCommand: BaseDeviceAPI<DeviceState, Command>['sendCommand']
+
+	private logger: StateHandlerContext['logger']
+
+	constructor(
+		private context: StateHandlerContext,
+		private config: StateHandlerConfig,
+		device: BaseDeviceAPI<DeviceState, Command>
+	) {
+		this.logger = context.logger
+
+		this.convertTimelineStateToDeviceState = (s, m) => device.convertTimelineStateToDeviceState(s, m)
+		this.diffDeviceStates = (o, n, m) => device.diffStates(o, n, m)
+		this.executeCommand = async (c) => device.sendCommand(c)
+
+		this.clock = setInterval(() => {
+			// main clock to check if next state needs to be sent out
+			const nextStates = this.stateQueue.filter((s) => s.state.time < Date.now() + 20) // any states to be executed before the next tick
+
+			nextStates.forEach((nextState) => {
+				const nextTime = Math.max(0, nextState?.state.time - Date.now())
+				setTimeout(() => {
+					if (!this._executingStateChange && this.stateQueue[0] === nextState) {
+						// if this is the next state, execute it
+						this.executeNextStateChange().catch((e) => {
+							this.logger.error('Error while executing next state change', e)
+						})
+					}
+				}, nextTime)
+			})
+		}, 20)
+	}
+
+	async terminate() {
+		clearInterval(this.clock)
+		this.stateQueue = []
+	}
+
+	async clearFutureStates() {
+		this.stateQueue = []
+	}
+
+	async handleState(state: Timeline.TimelineState<TSRTimelineContent>, mappings: Mappings) {
+		const nextState = this.stateQueue[0]
+
+		const trace = startTrace('device:convertTimelineStateToDeviceState', { deviceId: this.context.deviceId })
+		const deviceState = this.convertTimelineStateToDeviceState(state, mappings)
+		this.context.emitTimeTrace(endTrace(trace))
+
+		this.stateQueue = [
+			...this.stateQueue.filter((s) => s.state.time < state.time), // TODO - can we smuggle a little and execute something for the next frame?
+			{
+				deviceState,
+				state,
+				mappings,
+
+				measurement: new Measurement(state.time),
+			},
+		]
+
+		if (nextState !== this.stateQueue[0]) {
+			// the next state changed
+			if (nextState) nextState.commands = undefined
+			this.calculateNextStateChange().catch((e) => {
+				this.logger.error('Error while executing next state change', e)
+			})
+		}
+	}
+
+	async setCurrentState(state: DeviceState) {
+		this.currentState = {
+			commands: [],
+			deviceState: state,
+			state: this.currentState?.state || { time: Date.now(), layers: {}, nextEvents: [] },
+			mappings: this.currentState?.mappings || {},
+		}
+		await this.calculateNextStateChange()
+	}
+
+	clearFuture(t: number) {
```

Comment:
I think this naming is unclear. There is already a `clearFutureStates`, and this `clearFuture` is very different.
Perhaps `clearStatesBeforeTime`?